### PR TITLE
Adds support for using `Base.mapreducedim!` with AbstractField and AbstractOperations

### LINKED ISSUE #1422: Passing abstract operations (and `ComputedField`?) to `mapreduce`
We should look into using `mapreduce` as a way of computing reduced grid metrics and statistics using abstract operations. This could speed up `AveragedField` (and upcoming `SummedField`, `IntegratedField`, etc.) and make them lighter by not having to store a scratch space.

There are two methods for `mapreduce`:

```julia
mapreduce(f, op, itrs...; [init])
mapreduce(f, op, A::AbstractArray...; dims=:, [init])
```

The first one might be useful for abstract operations (as long as `getindex` is defined?).

Might be useful to do a bit of benchmarking to see if we can use them (and potentially get rid of some existing code).

The second function I guess would only be useful if we already have a `ComputedField` with an array.

We already use `mapreduce` a bit (the second method) but crucially we needed to pass in a view into a `CuArray` (and not a view into an `OffsetArray{CuArray}`): https://github.com/CliMA/Oceananigans.jl/blob/master/src/Solvers/preconditioned_conjugate_gradient_solver.jl#L57-L58

### PR DESCRIPTION
**Pull request by glwagner**
This PR adds support for in-place reduction operations where the target is an `AbstractReducedField` and the source is either an `AbstractDataField`, `AbstractOperation`, or array of some kind. These work on the GPU. In the end the solution was simple since we now subtype `AbstractArray`; we only need to pass a view into the interior indices of the target to `Base.mapreducedim!` (which on the GPU ends up at `GPUArrays.mapreducedim!`).

The result is that `AveragedField(op::AbstractOperation)` no longer needs to allocate memory for the three-dimensional result of computing `op`. Instead, `op` is reduced in a kernel. This is both faster (much much faster, I think --- though a benchmark is a good idea) and more memory efficient. It also greatly simplifies `compute!(field::AveragedField)`:

```julia
function compute!(avg::AveragedField, time=nothing)
    compute_at!(avg.operand, time)
    mean!(avg, operand)
    return nothing
end
```

So, resolves #1422.

I also took the liberty of resolving #1610 and nuking `interiorparent`...

### CODE REVIEW

File: src/Fields/averaged_field.jl [hunk 1]
```diff
@@ -56,27 +53,14 @@ function AveragedField(operand::AbstractField; dims, data=nothing, recompute_saf
...
-    i = operand_averaging_indices(Xa, Xo, topology(avg.grid, 1), avg.grid.Nx, avg.grid.Hx)
-    j = operand_averaging_indices(Ya, Yo, topology(avg.grid, 2), avg.grid.Ny, avg.grid.Hy)
-    k = operand_averaging_indices(Za, Zo, topology(avg.grid, 3), avg.grid.Nz, avg.grid.Hz)
-
-    operand_sans_halos = view(operand_parent, i, j, k)
-
-    mean!(avg.data.parent, operand_sans_halos)
-
+    avg .= 0 # should we have to do this?
+    mean!(avg, avg.operand)
```
Comment (tomchor) [inline src/Fields/averaged_field.jl:63]: Much simpler!

File: src/Forcings/continuous_forcing.jl [hunk 2]
```diff
@@ -127,8 +127,8 @@ Base.show(io::IO, forcing::ContinuousForcing{Nothing, Nothing, Nothing, P}) wher
         "├── parameters: $(forcing.parameters)", '\n',
         "└── field dependencies: $(forcing.field_dependencies)")
 
-Adapt.adapt_structure(to, forcing::ContinuousForcing{X, Y, Z}) where {X, Y, Z} =
-    ContinuousForcing{X, Y, Z}(Adapt.adapt(to, forcing.func),
+Adapt.adapt_structure(to, forcing::ContinuousForcing{LX, LY, LZ}) where {LX, LY, LZ} =
+    ContinuousForcing{LX, LY, LZ}(Adapt.adapt(to, forcing.func),
```
Comment (tomchor) [inline src/Forcings/continuous_forcing.jl:131]: Any specific reason why you changed `X` to `LX` throughout?

**Review by tomchor — APPROVED** [review]
Seems like a very positive change

File: src/BuoyancyModels/buoyancy_field.jl [hunk 3]
```diff
@@ -1,9 +1,11 @@
 using Adapt
 using KernelAbstractions
-using Oceananigans.Fields: AbstractDataField, FieldStatus, validate_field_data, new_data, conditional_compute!
+
+using Oceananigans.Fields: AbstractDataField, FieldStatus, validate_field_data, conditional_compute!
 using Oceananigans.Fields: architecture, tracernames
 using Oceananigans.Architectures: device
-using Oceananigans.Utils: work_layout
+using Oceananigans.Utils: work_layout, new_data
+using Oceananigans.Grids: new_data
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/BuoyancyModels/buoyancy_field.jl:8]: Not correct -- nice catch!

File: src/Forcings/continuous_forcing.jl (hunk 2 above)
Comment (glwagner) [reply src/Forcings/continuous_forcing.jl:131]: @ali-ramadhan prefers LX, etc (stands for "x location"). "X" could be confused with the x-coordinate, perhaps. It's pervasive throughout the code but I feel it never hurts to pay off technical debt when you have the chance. Sisyphus at work...

**Pull request merged**